src/pybind/mgr/mgr_util.py

```diff
@@ -335,6 +338,116 @@ def get_all_filesystems(self) -> List[str]:
+        try:
+            return json.loads(data.strip())
+        except json.JSONDecodeError:
+            logger.error('Failed to parse JSON for %s: %r', ctx, data)
+            return None
+
+    def _fetch_subvolume_groups(self, volume: str) -> List[str]:
+        """Fetch subvolumegroups for a given volume."""
+        cmd = {'prefix': 'fs subvolumegroup ls', 'vol_name': volume}
+        data = self._fetch_data(cmd, "subvolumegroups")
+        return [subvolgrp["name"] for subvolgrp in data] if data else []
+
+    def _fetch_subvolumes(self, volume: str, subvolumegrp: str) -> List[str]:
+        """Fetch subvolumes for a given volume and subvolumegroup."""
+        cmd = {
+            'prefix': 'fs subvolume ls',
+            'vol_name': volume,
+            'group_name': subvolumegrp if subvolumegrp != '_nogroup' else '',
+        }
+        data = self._fetch_data(cmd, "subvolumes")
+        return [subvol["name"] for subvol in data] if data else []
+
+    def _fetch_from_path(self, path: str, names: List[str], default: Optional[str] = None) -> Optional[str]:
+        """Extract a name from a path using the provided list of names."""
+        path_parts = path.split('/')
+        for part in path_parts:
+            if part in names or (default and part == default):
+                return part
+        return None
+
+    def _fetch_subvolumegroup_from_path(self, volume: str, path: str) -> Optional[str]:
+        """Extract subvolumegroup from a path."""
+        subvolumegroups = self._fetch_subvolume_groups(volume)
+        return self._fetch_from_path(path, subvolumegroups, '_nogroup')
+
+    def _fetch_subvolume_from_path(self, volume: str, path: str, subvolumegroup: Optional[str] = None) -> Optional[str]:
+        """Extract subvolume from a path."""
+        subvolumes = self._fetch_subvolumes(volume, subvolumegroup or '')
+        return self._fetch_from_path(path, subvolumes)
+
+    def get_earmark(
+        self,
+        path: str,
+        volume: str,
+        subvolgrp: Optional[str] = None,
+        subvolume: Optional[str] = None,
+    ) -> Optional[str]:
+        """
+        Given a volume, subvolumegroup, subvolume, and path, return the earmark.
+        """
+        if not subvolgrp:
+            subvolgrp = self._fetch_subvolumegroup_from_path(volume, path)
+        if not subvolume:
+            subvolume = self._fetch_subvolume_from_path(volume, path, subvolgrp)
+        cmd = {
+            'prefix': 'fs subvolume earmark get',
+            'vol_name': volume,
+            'group_name': subvolgrp if subvolgrp != '_nogroup' else '',
+            'sub_name': subvolume,
+        }
+        logger.debug('Getting earmark for subvolume: %r', cmd)
+        ret, data, status = self._mgr.mon_command(cmd)
+        if ret != 0:
+            logger.error('Failed to get earmark: %r', status)
+            return None
+        logger.info('Got earmark: %r', data)
+        return data.strip()
```

Comment:
I'm confused why'd we go through the mon_command interface for this. I think we should directly access the directory and the xattr in this code. Note: for smb (just smb, not nfs) we may eventually need to set other xattrs (acls) so we might as well get in the habit of working directly with the file system in this module.

Reply:
I preferred former approach as that's how I usually see in code elsewhere, but this makes sense to use xattrr directly for given `fs_handler`. I've updated the patch